lwm.man
.\" lwm, a window manager for X11
.\" Copyright (C) 1997-2016 Elliott Hughes, James Carter
.\" 
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
.\" 
.TH LWM 1
.SH NAME
lwm \- Lightweight Window Manager for the X Window System
.SH SYNTAX
\fBlwm \fP[\fB\-s\fP session-id] [\fB\-debugcli\fP] 
.SH DESCRIPTION
\fILwm\fP is a window manager for the X Window System.  It provides enough
features to allow the user to manage their windows, and no more.
.PP
Windows are surrounded by a frame with a titlebar at the top next to a close
icon (in the form of a cross).  The window with input focus is coloured
differently from the others.
.PP
\fILwm\fP uses the sloppy-focus scheme by default, whereby moving the pointer
into a window gives it input focus, but moving the pointer out does not cause
focus to be lost until the pointer enters another window.  Click-to-focus mode
can be enabled using the \fIfocus\fP resource.
.PP
A button 1 click on a window frame brings it to the top.  Dragging
button 1 on the frame of a resizable window repositions that edge of
the window.  If a corner rather than an edge is dragged, then both edges
forming the corner are repositioned.  While you're reshaping a window,
a little window pops up to show you the window's current size. Whether there is
a resize widget at the top of the title bar depends on Xresource config - see
below.
.PP
In the default configuration, button 1 on the root window does nothing.
.PP
Button 2 is used to drag a window by its frame, repositioning the window
but maintaining its position in the window stack.  This always just moves the
window, never resizing it.
.PP
In the default configuration, button 2 on the root window brings up a
new shell.
.PP
A button 3 click on a window frame hides that window.  Pressing
button 3 on the root window brings up a menu of all the windows.  Moving the
pointer over menu items while holding button 3 displays a coloured box
showing the window outline (even if it's hidden).
Releasing the button while over an item will unhide or raise the named window,
and give it focus.  The menu has two sections: the items above the dotted
line refer to hidden windows; those below to already visible windows.
.PP
A button 3 click in the frame while Shift is held down pushes the window
to the back, under any other windows.
.PP
A window may be closed by clicking on the little white cross in the top-left
corner of its frame.
.PP
It is possible to run commands when alt-clicking on a window's title bar.
The configured command is run with the window id (in hex with 0x prefix)
appended to it (separated by a space), so that the command run can be a script
that does some user-specified action to the window.
Separate commands can be configured for buttons 1 and 2. The default for
alt-button1 on the title bar is to run "force_title.sh", which can be found
alongside the LWM source code. This script allows the user to override a
window's visible title, so that annoying web apps that animate the window
title can be muted.
.SH OPTIONS
\fILwm\fP accepts the following command line options:
.PP
.TP 8
.B \-s
specifies a client ID for the X Session Management system, and is used
exclusively by session managers.
.TP 8
.B \-debugcli[=command1;command2...]
enables the debug CLI (command line interface), which allows LWM to be given
certain commands on stdin.  This is intended for debugging, not for normal LWM
use.  For details of commands available on the CLI, refer to the BUGS file.
If an '=commands' suffix is present, it is interpreted as a sequence of debugcli
commands, separated by ';', which LWM's debugcli will execute after startup is
complete.
.TP 8
.B \-test
runs all LWM's unit tests, exiting successfully if all tests pass.
.SH RESOURCES
\fILwm\fP understands the following X resources:
.TP 12
.B focus
if set to 'click', click-to-focus mode is activated, otherwise sloppy focus
.TP 12
.B titleFont
font used in window titles, and in the popup menu and size indicator
.TP 12
.B border
width in pixels of window borders
.TP 12
.B topBorder
how many of the top pixels of the title bar should be used as a resize widget.
Set this to zero to disable resizing on the title bar entirely (in which case,
the whole title bar, except the top-left and -right corners, will move the
window).
.TP 12
.B button1
program spawned when button 1 is clicked on the root window
.TP 12
.B button2
program spawned when button 2 is clicked on the root window
.TP 12
.B altButton1Title
program spawned when alt+button 1 is clicked on a window's title bar. The window
ID will be appended to this string (in hex with 0x prefix, separated with space)
and the resulting program will be run. Default value is "force_title.sh", which
for window 0x120002 will be run as "force_title.sh 0x120002".
.TP 12
.B altButton2Title
program spawned when alt+button 2 is clicked on a window's title bar. See the
description of altButton1Title for more info.
.TP 12
.B titleBGColour
background colour of the active window's title bar
.TP 12
.B borderColour
background colour of the active window's border
.TP 12
.B inactiveBorderColour
background colour of inactive windows' title bars and borders
.TP 12
.B windowHighlightColour
colour of the box showing window outlines from the window list menu
.TP 12
.B titleColour
colour of the title bar text in the active window
.TP 12
.B inactiveTitleColour
colour of the title bar text in inactive windows
.TP 12
.B closeIconColour
colour of the close icon in the active window
.TP 12
.B inactiveCloseIconColour
colour of the close icon in inactive windows
.TP 12
.B popupTextColour
text colour of the popup window (unhide menu and resizing popup)
.TP 12
.B popupBackgroundColour
background colour of the popup window
.TP 12
.B appIcon
where to show application icons. Possible values are 'none', 'title' (show only
in the app's window title bar), 'menu' (only in the unhide menu) or 'both'.
.TP 12
.B focusDelayMillis
how many milliseconds to delay repeated sloppy-focus focus events by. The
default is 50ms. If you find this is slow, reduce it. If you find that you can
move the mouse from window A to window B to window C, and sometimes have focus
remaining in window B, then increase this time a bit.
.SH "SEE ALSO"
.PP
X(7)
.SH AUTHORS
Elliott Hughes <ehughes@bluearc.com>,
James Carter <james@jfc.org.uk>,
Phil Norman <philnorm@gmail.com>


ChangeLog
Change Log for "lwm"


2020-07-04	pn	Basel
	Added a delay timer which kicks in when focus-follows-mouse changes the
	focus recipient too quickly. This can cause a race condition in X11,
	so that the wrong window ends up stealing focus.


2020-06-13	pn	Basel
	Added ability to run user-configured commands on an alt-click on the
	title bar of a window. By default (assuming the new force_title.sh and
	setvisname programs are on the user's path), alt-left-click on the title
	bar of a window will pop up a zenity dialog box, and allow the user to
	enter a new name for the window. This is then set as the
	_NET_WM_VISIBLE_NAME property on the window, which LWM will then display
	in preference to the client's own provided name.
	This is mainly so that annoying web apps which animate the window title
	can be muted. The 1Hz constant flapping that Google Hangouts Chat keeps
	up has been annoying the hell out of me for months now.


2020-05-22	pn	Basel
	Wrote 'ruler', a debug aid that displays a 2d ruler on the root window.
	Fixed a few bad window lookup bugs, which were dumping stack traces.


2020-05-19	pn	Basel
	Major refactor of how window sizing, movement is dealt with. Should fix
	a lot of little bugs.


2020-04-26	pn	Basel
	Moved geometry declarations from lwm.h to geometry.h.

	Added CreateNamedWindow function in xlib.(cc|h), which does the normal
	create window call but also assigns it a name. This means we can now
	see what each LWM window is for in 'xwininfo -root -tree'.


2020-04-25	pn	Basel
	Gave xlib.(cc|h) a namespace, and added there a bunch of shim functions
	which call same-named Xlib functions. Changed code around LWM to call
	through these new shim layers. The idea is to make it easier to track
	events sent to clients, and thus make it simpler to debug problems such
	as the disappearing ImageMagick windows.


2020-04-18	pn	Basel
	Deleted support for colormaps. 24bpp has been standard for decades.


2020-04-15	pn	Basel
	Was bitten by a Google Chrome update, which caused Chrome windows to
	drop all keyboard input for windows which weren't freshly opened.
	Figured out a fix for it, which still works for Java.


2019-09-15	pn	Basel
	Stopped 'withdrawn' windows from showing up in the unhide menu. This
	solves a bug where the rhythmbox preferences dialog would remain in the
	unhide menu even when closed, because the window was never destroyed,
	merely withdrawn (and reused again when next opened).


2019-09-01	pn	Basel
	Removed the poisonous 'PointerMotionHintMask', which was causing us to
	miss out on mouse motion events, and preventing the cursor from being
	properly updated according to which bit of the window furniture it was
	on. Also fixed a silly bug where we'd fail to switch to the
	'move window' cursor should the pointer move from the client window
	into the title bar.

	Got rid of the long-standing bug where we'd allow the window to be moved
	if the mouse pointer is on the furniture window's 1px border.


2019-08-29	pn	Basel
	Got rid of static functions to improve debugging.

	Added backtrace printing (for linux only so far) on X proto errors.

	Added reference counting and proper clean-ups of image icons for deleted
	windows. This fixes a pixmap leak, and also some bad over-cacheing
	behaviour where sometimes apps get each other's icons.


2019-05-25	pn	Basel
	Fixed sloppy divide-by-zero bug in xrandr window moving code.


2019-05-20	pn	Basel
	New algorithm for calculating where to position windows after a change
	of screen layout.

	This also introduces 'tests.cc', LWM's first ever unit tests! They're
	not using any standard library, and are compiled into the main binary
	(call by running 'lwm -test'). This isn't strictly proper unit testing,
	but it means I don't need to bother with a more complicated Makefile.
	And since LWM really isn't set up for easy unit testing (still having
	nearly everything in lwm.h), doing things properly would involve serious
	refactoring. Anyway, as far as I can tell, adding the tests in this way
	increases the binary size by 71KiB, which isn't going to hurt anyone.


2019-05-07	pn	Basel
	Fixed calculation of the title bar bounds, so that we don't have a
	little square at the right edge of the title bar, which is a resize
	widget when it should be a move one.

	Fixed the parsing of numeric Xresources so they don't treat all 0 values
	as being parse failures. This allows the top-of-window size-change area
	to be switched off (previously it would have fallen back to the
	default).


2019-05-05	pn	Basel
	Added back the ability to resize windows by dragging the top edge of the
	window frame. There's a new xresource called 'topBorder' which describes
	how many pixels thick this is (so it can be a different thickness to the
	normal window border); setting this to zero makes the whole title bar
	a move, not resize, widget. JFC likes resizing from the top, so this
	should make this branch of LWM more appealing.

	Added an 'appIcon' Xresource, which allows application icons to be
	enabled or disabled for title bars and the unhide menu.

	Fixed another little bug where LWM was moving menu's window on shutdown.
	This was caused by Client_MakeSane always subtracting textHeight() from
	the Y position, irrespective of whether the client is framed. The
	treatment of window sizes *really* needs more structure.


2019-05-03	pn	Basel
	Fixed another silly bug where strutted windows would be forcibly pushed
	out of their reserved areas.


2019-05-02	pn	Basel
	Fixed a silly bug where strutted windows would be forcibly pushed out of
	their own reserved areas on an xrandr screen change.


2019-05-01	pn	Basel
	Added 'xdbg.cc', a simple debug tool to measure locations on screen.
	Useful for debugging mouse positioning issues with LWM.

	Finally fixed LWM's issues with popup windows being opened outside
	the visible area of the screen (happens after a few laptop external
	screen add/remove sequences, with evergreen and terminator).


2019-02-27	pn	Basel

lwm.cc
/*
 * lwm, a window manager for X11
 * Copyright (C) 1997-2016 Elliott Hughes, James Carter
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include <errno.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>

#include <sys/timerfd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#include <signal.h>

#include "lwm.h"
#include "xlib.h"

bool is_initialising;
Display* dpy;  // The connection to the X server.

XftFont* g_font;
XftColor g_font_active_title;
XftColor g_font_inactive_title;
XftColor g_font_popup_colour;

bool shape;       // Does server have Shape Window extension?
int shape_event;  // ShapeEvent event type.

// Atoms we're interested in. See the ICCCM for more information.
Atom wm_state;
Atom wm_change_state;
Atom wm_protocols;
Atom wm_delete;
Atom wm_take_focus;
Atom compound_text;

// Netscape uses this to give information about the URL it's displaying.
Atom _mozilla_url;

// if we're really short of a clue we might look at motif hints, and
// we're not going to link with motif, so we'll have to do it by hand
Atom motif_wm_hints;

bool forceRestart;
char* argv0;

void rrScreenChangeNotify(XEvent* ev);
void setScreenAreasFromXRandR();

std::vector<std::string> Split(const std::string& in,
                               const std::string& split) {
  std::vector<std::string> res;
  int start = 0;
  while (true) {
    int end = in.find(split, start);
    if (end == std::string::npos) {
      res.push_back(in.substr(start));
      return res;
    }
    res.push_back(in.substr(start, end - start));
    start = end + split.size();
  }
}

/*ARGSUSED*/
extern int main(int argc, char* argv[]) {
  DebugCLI* debugCLI = nullptr;
  argv0 = argv[0];
  std::vector<std::string> debug_init_commands;
  for (int i = 1; i < argc; i++) {
    if (!strncmp(argv[i], "-debugcli", 9)) {
      debugCLI = new DebugCLI;
      if (argv[i][9] == '=') {
        // Argument is a sequence of commands, separated by ;.
        debug_init_commands = Split(std::string(argv[i] + 10), ";");
      }
    } else if (!strcmp(argv[i], "-test")) {
      LOGI() << "Run in self-test mode; will run all tests, then exit";
      return RunAllTests() ? 0 : 1;
    }
  }

  is_initialising = true;
  setlocale(LC_ALL, "");

  // Open a connection to the X server.
  dpy = XOpenDisplay(NULL);
  if (dpy == 0) {
    panic("can't open display.");
  }
  if (ScreenCount(dpy) != 1) {
    fprintf(stderr,
            "Sorry, LWM no longer supports multiple screens, and you "
            "have %d set up.\nPlease consider using xrandr.\n",
            ScreenCount(dpy));
  }
  Resources::Init();

  // Set up an error handler.
  XSetErrorHandler(errorHandler);

  // Set up signal handlers.
  signal(SIGTERM, Terminate);
  signal(SIGINT, Terminate);
  signal(SIGHUP, Terminate);

  // Ignore SIGCHLD.
  struct sigaction sa;
  sa.sa_handler = SIG_IGN;
#ifdef SA_NOCLDWAIT
  sa.sa_flags = SA_NOCLDWAIT;
#else
  sa.sa_flags = 0;
#endif
  sigemptyset(&sa.sa_mask);
  sigaction(SIGCHLD, &sa, 0);

  // Internalize useful atoms.
  wm_state = XInternAtom(dpy, "WM_STATE", false);
  wm_change_state = XInternAtom(dpy, "WM_CHANGE_STATE", false);
  wm_protocols = XInternAtom(dpy, "WM_PROTOCOLS", false);
  wm_delete = XInternAtom(dpy, "WM_DELETE_WINDOW", false);
  wm_take_focus = XInternAtom(dpy, "WM_TAKE_FOCUS", false);
  compound_text = XInternAtom(dpy, "COMPOUND_TEXT", false);
  _mozilla_url = XInternAtom(dpy, "_MOZILLA_URL", false);
  motif_wm_hints = XInternAtom(dpy, "_MOTIF_WM_HINTS", false);

  ewmh_init();

  int screenID = DefaultScreen(dpy);
  std::string titleFont = Resources::I->Get(Resources::TITLE_FONT);
  g_font = XftFontOpenName(dpy, screenID, titleFont.c_str());
  if (g_font == nullptr) {
    fprintf(stderr, "Couldn't find font %s; falling back", titleFont.c_str());
    g_font = XftFontOpenName(dpy, 0, "fixed");
    if (g_font == nullptr) {
      panic("Can't find a font");
    }
  }
  XRenderColor xrc = Resources::I->GetXRenderColor(Resources::TITLE_COLOUR);
  XftColorAllocValue(dpy, DefaultVisual(dpy, screenID),
                     DefaultColormap(dpy, screenID), &xrc,
                     &g_font_active_title);
  xrc = Resources::I->GetXRenderColor(Resources::INACTIVE_TITLE_COLOUR);
  XftColorAllocValue(dpy, DefaultVisual(dpy, screenID),
                     DefaultColormap(dpy, screenID), &xrc,
                     &g_font_inactive_title);
  xrc = Resources::I->GetXRenderColor(Resources::POPUP_TEXT_COLOUR);
  XftColorAllocValue(dpy, DefaultVisual(dpy, screenID),
                     DefaultColormap(dpy, screenID), &xrc,
                     &g_font_popup_colour);

  LScr::I = new LScr(dpy);
  LScr::I->Init();
  session_init(argc, argv);

  // Initialisation is finished; from now on, errors are not going to be fatal.
  is_initialising = false;

  // Do we need to support XRandR?
  int rr_event_base, rr_error_base;
  bool have_rr = XRRQueryExtension(dpy, &rr_event_base, &rr_error_base);
  if (have_rr) {
    XRRSelectInput(dpy, LScr::I->Root(), RRScreenChangeNotifyMask);
    setScreenAreasFromXRandR();
  }

  // See if the server has the Shape Window extension.
  shape = serverSupportsShapes();

  // The main event loop.
  int dpy_fd = ConnectionNumber(dpy);
  int max_fd = dpy_fd + 1;
  int delayed_focus_fd = LScr::I->GetFocuser()->GetTimerFD();
  if (ice_fd >= max_fd) {
    max_fd = ice_fd + 1;
  }
  if (delayed_focus_fd >= max_fd) {
    max_fd = delayed_focus_fd + 1;
  }

  // Just before we start the loop, execute any commands we've been told to
  // run on start-up.
  if (debugCLI) {
    debugCLI->Init(debug_init_commands);
  }

  while (!forceRestart) {
    fd_set readfds;

    FD_ZERO(&readfds);
    FD_SET(dpy_fd, &readfds);
    FD_SET(delayed_focus_fd, &readfds);
    if (ice_fd > 0) {
      FD_SET(ice_fd, &readfds);
    }
    if (debugCLI) {
      FD_SET(STDIN_FILENO, &readfds);
    }
    if (select(max_fd, &readfds, NULL, NULL, NULL) > -1) {
      if (FD_ISSET(dpy_fd, &readfds)) {
        while (XPending(dpy)) {
          XEvent ev;
          XNextEvent(dpy, &ev);
          // xrandr notifications have arbitrary numbers, so check for them
          // before trying the static selection.
          if (ev.type == rr_event_base + RRScreenChangeNotify) {
            rrScreenChangeNotify(&ev);
          } else {
            DispatchXEvent(&ev);
          }
        }
      }
      if (ice_fd > 0 && FD_ISSET(ice_fd, &readfds)) {
        session_process();
      }
      if (FD_ISSET(delayed_focus_fd, &readfds)) {
        LScr::I->GetFocuser()->TimerFDTriggered();
        // My best guess as to why we need this is that, because the event we
        // received didn't come off the Xlib connection (but rather our own
        // timer file descriptor), messages to the X server don't get flushed
        // automatically. The effect of this is that the delayed focus granting
        // sort of happens, but doesn't look like it did until the user does
        // something that triggers any event in LWM (like moving the mouse by
        // a pixel).
        // So call XSync so that we're sure all outstanding messages to, for
        // example, tell the client it has input focus, and redraw its frame,
        // get through.
        XSync(dpy, false);
      }
      if (debugCLI && FD_ISSET(STDIN_FILENO, &readfds)) {
        debugCLI->Read();
      }
    }
  }
  // Someone hit us with a SIGHUP: better exec ourselves to force a config
  // reload and cope with changing screen sizes.
  execvp(argv0, argv);
}

void rrScreenChangeNotify(XEvent* ev) {
  XRRScreenChangeNotifyEvent* rrev = (XRRScreenChangeNotifyEvent*)ev;
  int nScrWidth = rrev->width;
  int nScrHeight = rrev->height;
  // If my laptop is connected to a screen that is switched off, of I try
  // to switch to an external screen when none is connected, LWM gets this
  // event with a new size of 320x200. This forces all the windows to be
  // crushed into tiny little boxes, which is really annoying to repair
  // once I've got the external screen connected and X sorted out again.
  // A simple solution to this is to ignore any notifications smaller than
  // the smallest vaguely sensible size I can think of (and, TBH, this is
  // really too small to be sensible already).
  if (nScrWidth < 600 || nScrHeight < 400) {
    LOGW() << "Ignoring tiny screen dimensions from xrandr: " << nScrWidth
           << "x" << nScrHeight;
    return;
  }

  static long lastSerial;
  if (rrev->serial == lastSerial) {
    LOGI() << "Dropping duplicate event for serial " << lastSerial;
    return;  // Drop duplicate message (we get lots of these).
  }
  lastSerial = rrev->serial;
  setScreenAreasFromXRandR();
}

void setScreenAreasFromXRandR() {
  XRRScreenResources* res = XRRGetScreenResourcesCurrent(dpy, LScr::I->Root());
  if (!res) {
    LOGE() << "Failed to get XRRScreenResources";
    return;
  }
  xlib::XFreer res_freer((void*)res);
  if (!res->ncrtc) {
    LOGE() << "Empty list of CRTs";
    return;
  }
  // Ignore any CRT with mode==0.
  // Change nScrWidth/nScrHeight according to the total extent of all visible
  // areas, and don't rely on the size provided in the event itself. This is
  // because when switching from internal+external monitors to internal only,
  // the first couple of notifications claim the old area. However, querying
  // the CRT info already gets the correct sizes and locations (including
  // mode=0 for those that are disabled).
  std::vector<Rect> visible;
  for (int i = 0; i < res->ncrtc; i++) {
    const RRCrtc crt = res->crtcs[i];
    LOGI() << "Looking up CRT " << i << ": " << crt;
    XRRCrtcInfo* crtInfo = XRRGetCrtcInfo(dpy, res, crt);
    LOGI() << "  CRT size " << crtInfo->width << "x" << crtInfo->height
           << ", offset " << crtInfo->x << "," << crtInfo->y
           << " (mode=" << crtInfo->mode << ")";
    if (!crtInfo->mode) {
      continue;
    }
    const int xMin = crtInfo->x;
    const int yMin = crtInfo->y;
    const int xMax = xMin + crtInfo->width;
    const int yMax = yMin + crtInfo->height;
    visible.push_back(Rect{xMin, yMin, xMax, yMax});
  }
  LScr::I->SetVisibleAreas(visible);
}

extern void RunCommand(const std::string& command) {
  const char* sh = getenv("SHELL");
  if (!sh) {
    sh = "/bin/sh";
  }
  const char* display_str = DisplayString(dpy);

  switch (fork()) {
    case 0:  // Child.
      close(ConnectionNumber(dpy));
      if (display_str) {
        const int len = strlen(display_str) + 9;
        char* str = (char*)malloc(len);
        snprintf(str, len, "DISPLAY=%s", display_str);
        putenv(str);
      }
      execl(sh, sh, "-c", command.c_str(), NULL);
      fprintf(stderr, "%s: can't exec \"%s -c %s\"\n", argv0, sh,
              command.c_str());
      execlp("xterm", "xterm", NULL);
      exit(EXIT_FAILURE);
    case -1:  // Error.
      fprintf(stderr, "%s: couldn't fork\n", argv0);
      break;
  }
}

extern void shell(int button) {
  std::string command;
  if (button == Button1) {
    command = Resources::I->Get(Resources::BUTTON1_COMMAND);
  } else if (button == Button2) {
    command = Resources::I->Get(Resources::BUTTON2_COMMAND);
  }
  if (command.empty()) {
    return;
  }
  RunCommand(command);
}

extern int textHeight() {
  return g_font->height;
}

extern void drawString(Window w,
                       int x,
                       int y,
                       const std::string& s,
                       XftColor* c) {
  int screenID = DefaultScreen(dpy);
  XftDraw* draw = XftDrawCreate(dpy, w, DefaultVisual(dpy, screenID),
                                DefaultColormap(dpy, screenID));
  XftDrawStringUtf8(draw, c, g_font, x, y,
                    reinterpret_cast<const FcChar8*>(s.c_str()), s.size());
  XftDrawDestroy(draw);
}

// Returns the width of the given string in pixels, rendered in the LWM font.
extern int textWidth(const std::string& s) {
  XGlyphInfo extents;
  XftTextExtentsUtf8(dpy, g_font, reinterpret_cast<const FcChar8*>(s.c_str()),
                     s.size(), &extents);
  return extents.xOff;
}
